Bump Debian version from Buster to Bullseye, and rebuild Linux libraries

illwieckz opened the pull request:
- Bump Debian version from Buster to Bullseye for Linux engine build.
- Bump Debian version from Buster to Trixie for things not producing linux binaries, like MinGW and NaCl vms.
- Bump CMake in Debian.
- Bump CMake in Darling.
- Rebuild as much as Linux libraries as possible.
- It now use three images for building engines: one for building Linux binaries, one for MinGW binaries, and one for macOS binaries, allowing finer configuration.
- Those images are now kept the same and aren't rebuilt according to the amount and nature of engine targets to be built.
- A rework preventing more image invalidations was done.

We want to use most recent libraries as possible, and those being the same on all platforms as much as we can.

It assumes DaemonEngine/Daemon#1433 is merged (so it should be merged first):

- https://github.com/DaemonEngine/Daemon/pull/1433

It obsoletes #36:

- https://github.com/Unvanquished/release-scripts/pull/36

illwieckz committed:
docker: bump Debian version from Buster to Bullseye

illwieckz committed:
docker: rebuild as much as Linux libraries as possible

We want to use most recent libraries as possible,
and those being the same on all platforms as much
as we can.

illwieckz commented:
About the Debian bump:

- https://github.com/Unvanquished/Unvanquished/issues/3393

For rebuilding as much as Linux libraries as we can, this is because it is stupid to rely on old-old-stable or old-stable libraries while we actually review newer libraries.

And there is no reason Linux builds get those very old libraries and don't benefit from newer optimizations and things like that while macOS and Windows get all the shiny toys.

illwieckz commented:
The reason why this still uses the system's nettle is because of that:

```
/Unvanquished/daemon/external_deps/linux-arm64-default_11/lib/libnettle.a(fat-arm64.o):
 in function `fat_init':
/Unvanquished/daemon/external_deps/build-linux-arm64-default_11/nettle/nettle-3.10.2/fat-arm64.c:289:
 (.text.startup+0x27c): relocation truncated to fit: R_AARCH64_LD64_GOTPAGE_LO15 against symbol
 `_nettle_sha256_compress_n_c' defined in .text section
 in /Unvanquished/daemon/external_deps/linux-arm64-default_11/lib/libnettle.a(sha256-compress-n.o)
/usr/lib/gcc-cross/aarch64-linux-gnu/10/../../../../aarch64-linux-gnu/bin/ld:
 /Unvanquished/daemon/external_deps/build-linux-arm64-default_11/nettle/nettle-3.10.2/fat-arm64.c:289:
 warning: too many GOT entries for -fpic, please recompile with -fPIC
collect2: error: ld returned 1 exit status
```

I'm aware of this for months now, maybe a year, but I never found what was causing this.

Actually this is not affecting amd64/i686 builds. This may be a bug in nettle.

So for now the Linux release build still relies on nettle as provided by the system.

slipher commented on docker/build-external-dependencies:
```diff
@@ -30,29 +30,22 @@ cd '/Unvanquished/daemon/external_deps'
 
 for target in ${targets}
 do
+	# For other architectures, tThe downloadable deps archive contains
```
Typo `tThe`

illwieckz replied:
Fixed.

slipher commented on docker/build-external-dependencies:
```diff
@@ -30,29 +30,22 @@ cd '/Unvanquished/daemon/external_deps'
-		# static libs are not multiarch compatible.
-		# webp: We want to make sure we use the version we want to use.
-		_exec ./build.sh "${target}-default" ogg opus opusfile curl sdl2 glew webp openal ncurses naclsdk install
+		_exec ./build.sh "${target}-default" release install
```
What's this `release` thing? You mention requiring https://github.com/DaemonEngine/Daemon/pull/1433 which changes the deps build script, but I spent many hours reviewing that and don't recall any target called `release`.

illwieckz replied:
See:

- https://github.com/DaemonEngine/Daemon/pull/1433/files#r2225911538

illwieckz replied:
I added it some days ago.

illwieckz replied:
This was reverted.

slipher commented on docker/install-system-dependencies:
```diff
@@ -83,6 +78,7 @@ linux_generic_av_packages=(
 	libasound2-dev:native
 	libaudio-dev:native
 	libgl1-mesa-dev:native
+ 	libglu1-mesa-dev:native
```
This is not a big deal, but we shouldn't need GLU. Previously I had to add some define like `GLEW_NO_GLU` to prevent a dependency header from trying to include it.

illwieckz replied:
Ah, I'll look for that define then. I had to add this because it didn't built without, but if we can disable it it's better.

illwieckz replied:
Done.

illwieckz committed:
docker: some rewrite

illwieckz commented:
I'm now switching to Bookworm as Debian old-stable since Trixie is out.

_Edit:_ I went back to Bullseye for Linux builds.

illwieckz committed:
build-release: build against our own Freetype

illwieckz committed:
docker: silence a warning when fixing ownership

Silence this warning:

> chown -R 1000.1000 /Unvanquished/build/release
> chown: warning: '.' should be ':': '1000.1000'

illwieckz committed:
build-release: bump macOS deployment target because OpenAL bumped it

illwieckz committed:
docker: install PipeWire headers to enable and build the PipeWire backend in OpenAL

illwieckz committed:
docker: install linux CMake from tarball

illwieckz committed:
docker: rewrite docker/install-xcode-and-packages

illwieckz committed:
docker: rewrite more darling stuff

illwieckz committed:
build-release: all mingw-i686 platforms now use the DWARF exception flavor

illwieckz committed:
docker: bump macOS CMake

illwieckz committed:
build-release: switch to SDL3

illwieckz committed:
docker: build MinGW engine with its own image running a newer Debian

illwieckz committed:
docker: use target-independent per-system images

illwieckz committed:
docker: add libraries to enable Wayland and other things on SDL3

illwieckz committed:
build-release: make linux-amd64 compiler explicit

illwieckz committed:
docker: bump to trixie when not producing native linux binaries

illwieckz committed:
docker: enable backports

illwieckz committed:
docker: we can avoid libglu1

illwieckz commented:
So https://github.com/DaemonEngine/Daemon/pull/1433 has been merged:

- https://github.com/DaemonEngine/Daemon/pull/1433

So we now need this branch to build the current Dæmon's `master` branch.

slipher commented on docker/build-external-dependencies:
```diff
@@ -30,29 +30,22 @@ cd '/Unvanquished/daemon/external_deps'
+		# We now rebuild as much as libraries as possible so we get the
+		# updated libraries on every platform and we get what we expect.
+		# Also, some libraries may require us to rebuild them.
+		# sdl3: Libraries are configured with unneeded dynamic depencencies.
```
typo `depencencies`

illwieckz replied:
Fixed.

slipher reviewed (commented):
I have reviewed `build-release` and `build-external-dependencies`.

LGTM modulo comment

illwieckz commented:
> I have reviewed `build-release` and `build-external-dependencies`.

Actually I forgot to remove the line building the deps for the VM (to fetch the nacl sdk) as we now rely on the dowloaded deps archive. This fix is just clean-up because that `build-external-dependencies` script is already not called anymore for vm builds, so that was just leftover I cleaned-up.

> LGTM modulo comment

Thanks.
